libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Analytics/SFSDKSalesforceAnalyticsManager.m

```diff
@@ -154,40 +154,46 @@ - (void) publishEvents:(NSArray<SFSDKInstrumentationEvent *> *) events {
-        BOOL success = YES;
-        NSArray<Class<SFSDKTransform>> *remoteKeySet = [self.remotes allKeys];
-        for (Class<SFSDKTransform> transformClass in remoteKeySet) {
-            if (transformClass) {
-                NSMutableArray<NSDictionary *> *eventsJSONArray = [[NSMutableArray alloc] init];
-                for (SFSDKInstrumentationEvent *event in events) {
-                    [eventIds addObject:event.eventId];
-                    NSDictionary *eventJSON = [transformClass transform:event];
-                    if (eventJSON) {
-                        [eventsJSONArray addObject:eventJSON];
-                    }
-                }
-                Class<SFSDKAnalyticsPublisher> networkPublisher = self.remotes[transformClass];
-                if (networkPublisher) {
-                    BOOL networkSuccess = [networkPublisher publish:eventsJSONArray];
-                    
-                    /*
-                     * Updates the success flag only if all previous requests have been
-                     * successful. This ensures that the operation is marked success only
-                     * if all publishers are successful.
-                     */
-                    if (success) {
-                        success = networkSuccess;
-                    }
+        for (SFSDKInstrumentationEvent *event in events) {
+            [eventIds addObject:event.eventId];
+        }
+        __block BOOL overallSuccess = YES;
+        __block BOOL overallCompletionStatus = NO;
+        __block NSMutableArray<Class<SFSDKTransform>> *remoteKeySet = [[self.remotes allKeys] mutableCopy];
+        __block Class<SFSDKTransform> curTransform = [remoteKeySet objectAtIndex:0];
+        PublishCompleteBlock publishCompleteBlock = ^void(BOOL success) {
+
+            /*
+             * Updates the success flag only if all previous requests have been
+             * successful. This ensures that the operation is marked success only
+             * if all publishers are successful.
+             */
+            if (success) {
+                overallSuccess = success;
+            }
+
+            // Removes current transform from the list since it's done.
+            if (remoteKeySet) {
+                [remoteKeySet removeObject:curTransform];
+            }
+
+            // If there are no transforms left, we're done here.
+            if (!remoteKeySet || remoteKeySet.count == 0) {
+                overallCompletionStatus = YES;
+            }
+            if (!overallCompletionStatus) {
+                curTransform = [remoteKeySet objectAtIndex:0];
+                [self applyTransformAndPublish:curTransform events:events publishCompleteBlock:publishCompleteBlock];
+            } else {
+
+                /*
+                 * Deletes events from the event store if the network publishing was successful.
+                 */
+                if (overallSuccess) {
+                    [self.eventStoreManager deleteEvents:eventIds];
```

Comment:
Events are batched and sent as a single request, so either the request was successful or it failed. However, since we support multiple endpoints, it's possible that one request was successful but another failed. In this case, we will not delete events. They might be sent twice to a previously successful endpoint, but that's acceptable as opposed to loss of data on another endpoint/service.